python/pynq/interrupt_map.py

```diff
@@ -0,0 +1,139 @@
+import re
+
+
+class InterruptMap:
+    """Class to extract interrupt information from a TCL configuration file
+
+    Attributes
+    ----------
+    intc_parent : dict [str, str, int]
+        All AXI interrupt controllers in the system attached to
+        a PS7 interrupt line. Key is the name of the controller and
+        value is parent interrupt controller and the line interrupt used
+        The PS7 is the root of the hierarchy and is unnamed
+
+    intc_pins : dict [str, str, int]
+        All pins in the design attached to an interrupt controller listed in
+        in intc_names. Key is the name of the pin, value is the interrupt
+        controller and line used.
+
+    """
+
+    def __init__(self, tcl_name):
+        """Returns a map built from the supplied tcl file
+
+        Paramters
+        ---------
+        tcl_name : str
+            The tcl filename to parse. This is opened directly so should be
+            fully qualified
+        """
+
+        if not isinstance(tcl_name, str):
+            raise TypeError("tcl_name has to be a string")
+
+        # TODO: support nested hierarchies
+
+        # Initialize result variables
+        self.intc_names = []
+        self.intc_parent = {}
+        self.concat_cells = {}
+        self.nets = []
+        self.pins = {}
+        self.intc_pins = {}
+        self.ps7_name = ""
+
+        # Key strings to search for in the TCL file
```

Comment:
The following string expression definitions seem too many. I would remove some of those and just use regex. For example, `create_bd_cell`. And I would not define expressions like "] $".